### ISSUE TITLE
[Feature Request][PyTorch] Deterministic and predictable behaviour of batch sampler 

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/pytorch#7209]
Today `DataLoader` works well and returns reproducible results when fixing `torch.manual_seed`.

However, the problem is that the batch sampler or sampler (if it uses torch random generator) returns differents results if it used alone or in data loader (even fixing the random state) when `num_workers > 0`

The problem is [here](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L261), when `base_seed` is randomly set, random indices generated by the first call of `next(sampler_iter)` that happens [after](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L296) are not the same as they'd have been generated alone (with `next`) or by the data loader with `num_workers=0`.

Can we set `base_seed` differently without using torch random generator?

For example, we can use a fix `base_seed` in `DataLoaderIter`:
```python
            base_seed = 7056
            self.workers = [
                multiprocessing.Process(
                    target=_worker_loop,
                    args=(self.dataset, self.index_queue, self.worker_result_queue, self.collate_fn,
                          base_seed + i, self.worker_init_fn, i))
                for i in range(self.num_workers)]
```
and set a random seed in `_worker_loop`:
```python
def _worker_loop(dataset, index_queue, data_queue, collate_fn, seed, init_fn, worker_id):
    # ...    
    base_seed = torch.LongTensor(1).random_(seed)[0]
    torch.set_num_threads(1)
    torch.manual_seed(base_seed)
    # ...
```
What do you think ?

Context: this can be helpful if I would like to resume `DataLoader` from a given iteration (if training has crashed for example).

Thanks

**Comment by SsnL** [comment]
I'm not sure that I understand. The data loading result today is already deterministic even when `num_workers > 0`. The `base_seed` is generated using the PyTorch random seed, which you may set before data loading.

Furthermore, your `base_seed = torch.LongTensor(1).random_(seed)[0]` will just generate same seed in every worker.

**Comment by SsnL** [comment]
And I don't think there is anything we can do to remain deterministic when resuming after crash. You either have to dump the RNG state after the crash, or do all the random generation for previous steps again.

**Comment by vfdev-5** [comment]
@SsnL sorry for being unclear, let me be more explicit. Yes, I confirm that today `DataLoader` is deterministic for `num_workers > 0` and the following code (pytorch 0.4.0) : 
```python
x = np.arange(30)

def transform(v):    return v + torch.rand(1).item()
        
class TestDataset(Dataset):    
    def __len__(self):        return 30    
    def __getitem__(self, index):        return transform(x[index]), x[index]
    
dataset = TestDataset()
torch.manual_seed(12345)
loader = DataLoader(dataset, batch_size=4, shuffle=True, num_workers=4, drop_last=True)

for batch in loader:
    print(batch)
```
gives every time the same results:
```
[tensor([ 15.9327,  17.3807,  10.6407,  28.2207], dtype=torch.float64), tensor([ 15,  17,  10,  28])]
[tensor([ 12.5148,  13.5664,  27.7817,   6.7614], dtype=torch.float64), tensor([ 12,  13,  27,   6])]
[tensor([ 25.5011,   8.8289,  16.3753,  18.9677], dtype=torch.float64), tensor([ 25,   8,  16,  18])]
[tensor([ 29.1541,   4.7404,  23.7058,   7.0936], dtype=torch.float64), tensor([ 29,   4,  23,   7])]
[tensor([ 19.2820,   9.9384,  22.6470,  20.5152], dtype=torch.float64), tensor([ 19,   9,  22,  20])]
[tensor([ 21.4232,   5.6175,   0.6057,  24.5759], dtype=torch.float64), tensor([ 21,   5,   0,  24])]
[tensor([ 11.7328,   2.0943,   3.6006,  14.7609], dtype=torch.float64), tensor([ 11,   2,   3,  14])]
```
And this is good. Pay attention on `target` values: 15, 17, 10, 28 etc. these values are generated by the batch sampler (in my case by `RandomSampler` with `BatchSampler`).

If I execute the previous code with `num_workers=0`, the result is: 
```
[tensor([  0.6659,  20.0413,  19.5361,  15.3120], dtype=torch.float64), tensor([  0,  20,  19,  15])]
[tensor([ 22.4288,  17.2770,  28.4377,  18.9498], dtype=torch.float64), tensor([ 22,  17,  28,  18])]
[tensor([ 21.0193,  26.2634,  11.9249,  16.4737], dtype=torch.float64), tensor([ 21,  26,  11,  16])]
[tensor([ 10.3961,   1.6681,  25.2888,  12.7053], dtype=torch.float64), tensor([ 10,   1,  25,  12])]
[tensor([  5.1916,  14.3497,  29.8827,  23.5490], dtype=torch.float64), tensor([  5,  14,  29,  23])]
[tensor([ 13.1045,   9.8004,   2.1677,  24.4677], dtype=torch.float64), tensor([ 13,   9,   2,  24])]
[tensor([  3.4986,   8.2797,  27.7238,   4.4506], dtype=torch.float64), tensor([  3,   8,  27,   4])]
```
and target values: 0, 20, 19, 15 etc are the same as if I run only:
```
torch.manual_seed(12345)
loader = DataLoader(dataset, batch_size=4, shuffle=True, num_workers=0, drop_last=True)
batch_sampler_iter = iter(loader.batch_sampler)
for indices in batch_sampler_iter:
    print(indices)
```
the output is 
```
[0, 20, 19, 15]
[22, 17, 28, 18]
[21, 26, 11, 16]
[10, 1, 25, 12]
[5, 14, 29, 23]
[13, 9, 2, 24]
[3, 8, 27, 4]
```
This difference is due to the fact that `RandomSampler` uses torch random generator on `next()` call. 
And in case of `num_workers>0` torch random generator is called once to generate `base_seed` before calling `next()`. 

*Why it can be important to reproduce the same indices?* I imagine an epoch with a large number of iterations, so that resuming crashed DataLoader state from a given iteration instead of epoch begining can be useful. So, to skip previous batches can be simply run multiple times `next` on the batch sampler.

Concerning the fix I propose, if I understand correctly `base_seed` was set as random number to avoid same random state in all workers and thus producing the same random augmentation.

> Furthermore, your `base_seed = torch.LongTensor(1).random_(seed)[0]` will just generate same seed in every worker.

Actually `seed` used to generate `base_seed` in the worker is setup as previously `base_seed + worker_id`, so they will be different for a single run and same for different epochs.

@SsnL hope being more clear now. What do you think about this ?

**Comment by apaszke** [comment]
That makes sense, thanks a lot for the detailed descriptions @vfdev-5. It seems to me that it would be sufficient to always take the `base_seed`, even if it's going to be unused (since it's really cheap to do so). Could you please send a PR with the fix?

**Comment by SsnL** [comment]
@vfdev-5 Thanks for explaining! I think @apaszke 's suggestion makes sense. Do you want to submit a PR?

**Comment by vfdev-5** [comment]
@SsnL @apaszke yes, I would like to submit a PR, with pleasure :)

**Comment by vfdev-5** [comment]
@apaszke a little more precision on your suggestion. Do you mean to instantiate `base_seed` outside the `num_workers` condition:
```python
class _DataLoaderIter(object):
    r"""Iterates once over the DataLoader's dataset, as specified by the sampler"""

    def __init__(self, loader):
        self.dataset = loader.dataset
        self.collate_fn = loader.collate_fn
        self.batch_sampler = loader.batch_sampler
        self.num_workers = loader.num_workers
        self.pin_memory = loader.pin_memory and torch.cuda.is_available()
        self.timeout = loader.timeout
        self.done_event = threading.Event()

        self.sample_iter = iter(self.batch_sampler)

        # base seed is randomly instantiated here
        base_seed = torch.LongTensor(1).random_()[0]
        if self.num_workers > 0:
            # ....
            # removed ouside -> base_seed = torch.LongTensor(1).random_()[0]
```

if yes, sure that this solves the problem of different batch indices for all values of `num_workers`, however indices still remain different of those generated by batch_sampler standalone:
```python
torch.manual_seed(12345)
loader = DataLoader(dataset, batch_size=4, shuffle=True, num_workers=any_value, drop_last=True)
for batch in loader:
    print(batch)
```
with the output:
```
[tensor([ 15.9327,  17.3807,  10.6407,  28.2207], dtype=torch.float64), tensor([ 15,  17,  10,  28])]
[tensor([ 12.5148,  13.5664,  27.7817,   6.7614], dtype=torch.float64), tensor([ 12,  13,  27,   6])]
[tensor([ 25.5011,   8.8289,  16.3753,  18.9677], dtype=torch.float64), tensor([ 25,   8,  16,  18])]
[tensor([ 29.1541,   4.7404,  23.7058,   7.0936], dtype=torch.float64), tensor([ 29,   4,  23,   7])]
[tensor([ 19.2820,   9.9384,  22.6470,  20.5152], dtype=torch.float64), tensor([ 19,   9,  22,  20])]
[tensor([ 21.4232,   5.6175,   0.6057,  24.5759], dtype=torch.float64), tensor([ 21,   5,   0,  24])]
[tensor([ 11.7328,   2.0943,   3.6006,  14.7609], dtype=torch.float64), tensor([ 11,   2,   3,  14])]
```
vs 
```python
torch.manual_seed(12345)
loader = DataLoader(dataset, batch_size=4, shuffle=True, num_workers=any_value, drop_last=True)
batch_sampler_iter = iter(loader.batch_sampler)
for indices in batch_sampler_iter:
    print(indices)
```
with the output:
```
[0, 20, 19, 15]
[22, 17, 28, 18]
[21, 26, 11, 16]
[10, 1, 25, 12]
[5, 14, 29, 23]
[13, 9, 2, 24]
[3, 8, 27, 4]
```

The problem is that when indices are "released" by
```python
            # prime the prefetch loop
            for _ in range(2 * self.num_workers):
                self._put_indices()
```
the random state is not the same anymore when set by `torch.manual_seed`
 
Another fix is to save to random state before initialization of `base_seed` and restore it just before 
`self._put_indices()`. 

@SsnL @apaszke what do you guys think about ?

**Comment by vfdev-5** [comment]
More on the first proposed solution to instantiate random seed in the worker, see the [diff](https://github.com/pytorch/pytorch/compare/master...vfdev-5:issue_7209?expand=1)

**Comment by apaszke** [comment]
I think that's ok. You can't assume that calling library code won't advance the RNGs. It's a really simple fix so I'd go with that.

**Comment by fmassa** [comment]
Fixed via https://github.com/pytorch/pytorch/pull/7265
